Docs (#11)
Observatory-Sciences/aravis-detector: docs/user/tutorials/tools.md

# Python Tools

As of version 0.0.1 prerelease the only available python tool is the Client Line Interface.

## ArvCLI

> Aravis CLI allows the user to control the frame processor through the use of the command line interface. It also provides the user with an example of how to send http requests to the odin server. It can be installed using ```pip install python/tools -e``` from the aravis detector directory.

Reviewer: ````suggestion
Aravis CLI allows the user to control the frame processor through the use of the command line interface. It also provides the user with an example of how to send http requests to the odin server. It can be installed into the virtualenv from the aravis detector directory. The environment should again be setup as described in the Run Tutorial:

```shell
export PATH=$MAIN_DIR/prefix/bin:$PATH
source $MAIN_DIR/venv/bin/activate
```

Then ```arvcli``` can be installed with: ```pip install python/tools -e```
````

        'frames_max': 1000,
        'frames_written': 1000,
        'frames_processed': 1000,
        'file_path': 'odin_camera_driver/temp',
        'file_name': 'run_6s48m15h17d4m2024y_000000.h5',
        'acquisition_id': ' ',
        'processes': 1,
        'rank': 0,
        'timeout_active': False,
        'timing': {
            'last_create': 681,
            'max_create': 681,
            'mean_create': 530,
            'last_write': 26141,
            'max_write': 35680,
            'mean_write': 25260,
            'last_flush': 41,
            'max_flush': 66,
            'mean_flush': 40,
            'last_close': 64,
            'max_close': 246,
            'mean_close': 155,
            'last_process': 1,
            'max_process': 35801,
            'mean_process': 1
        }
    },
    'view': {'timing': {'last_process': 11, 'max_process': 43439, 'mean_process': 13}},
    'error': [
        'Frame invalid',
        'Frame destined for [data] but dataset has not been defined in the HDF plugin'
    ],
    'timestamp': '2024-04-18T16:30:11.005675',
    'connected': True
}
```

To get more specific values the user can call the ```--get``` command.

```shell

> $arvcli --get fp/config/aravis/frame_rate

Reviewer: ```suggestion
arvcli --get fp/config/aravis/frame_rate
```

Either remove the `$` or make a space between it and the command to signify that this is in a shell - although you haven't done that in other example shell snippets so seems a little bit out of place to start that here...

frame_rate = 5
```

To change more specific values the user can call the ```--put``` command.

```shell

> $arvcli --put fp/config/aravis/frame_rate

Reviewer: ```suggestion
arvcli --put fp/config/aravis/frame_rate
```

#### Get requests

Any paths starting with the "fp" value are truncated after the 3rd value. The first part is added as part of the http request and the later parts are used to parse the json file.

In the above case, the "fp/config/aravis/" is transformed into "http:/< ip_address >/api/0.1/fp/config/aravis/" and the value frame_rate is used as a json key. "sys" paths are split immediately after sys, and aravis paths are sent as http paths directly.

#### Put request

All put requests are either of the form "aravis/config" or "fp/config/< plugin >/" with one more path key. The additional key is then used to create a json file sent to the server together with the input variable.

#### Valid path list

The following is a list of example tree of paths that work with the aravis, live view and hdf plugin for the get function:

```shell
sys/
fp/ status  / aravis
            / view
            / hdf
    config  / aravis
            / view
            / hdf
aravis  / status
        / config
```

To each of this paths an additional value key can be added (eg. "frame_rate") to target a specific value. To get an accurate list of possible simply call the path as follows:

```shell

> $arvcli --get fp/config/aravis

Reviewer: ```suggestion
arvcli --get fp/config/aravis
```

aravis = {'ip_address': '127.0.0.1', 'camera_id': '', 'camera_serial_number': '', 'camera_model': '',
        'exposure_time': 0.0, 'frame_rate': 0.0, 'frame_count': 1500, 'pixel_format': 'Mono8', 
        'acquisition_mode': 'Continuous', 'status_frequency_ms': 1000, 'empty_buffers': 50,
        'file_path': 'odin_camera_driver/temp', 'data_set_name': 'data', 'file_name': 'test'}
```

### Commands

> Arvcli has subcommands (eg, "--get", "--put") that take one argument and commands. Commands are more complex and have their own subcommands.

Reviewer 2: These double minus "--" are not showing up properly when generated.  Perhaps you can escape them or make them mini code blocks etc.

Author: Changed to code block

To get a list of subcommands for a command simply use the `--help` tag after stating the command as follows:

```shell

> $arvcli connect --help

Reviewer: ```suggestion
arvcli connect --help
```

> $arvcli connect --help

Reviewer 2: Same here, double minus not showing up properly

                                                                                   
╭─ Options ───────────────────────────────────────────────────────────────────────╮
│ --start  -s                start file writer                                    │
│ --stop   -s                stop the file writer                                 │
│ --write  -w                start file writer and start acquisition              │
│ --stopW  -sw               stop file writer and acquisition                     │
│ --file   -f       TEXT     saving file name [default: None]                     │
│ --path   -p       TEXT     path to the directory [default: None]                │
│ --num    -n       INTEGER  number of files to save [default: None]              │
│ --help                     Show this message and exit.                          │
╰─────────────────────────────────────────────────────────────────────────────────╯
```

The file name defaults to a string value of the form "run_<time stamp\>" and the path and file number default to the values specified in the config file.

### Settings and custom commands

> Arvcli uses a config.yml file to store the IP address and port of the odin-control server as well as default values. This file can be edited directly (and this is encouraged) but the user can also change the IP address and port using the ```--ip``` and ```-port``` commands.

Reviewer 2: These double minus signs worked.  Apply the same formatting to those above :+1:
